src/lib/autodiscovery/autodiscovery_task_executor.py

```diff
@@ -0,0 +1,108 @@
+        )
+        await autodiscovery_task.get_task_result()
+        return autodiscovery_task
+
+    def __init__(
+        self,
+        services: List[GCPService],
+        autodiscovery_manager: AutodiscoveryManager,
+        current_extension_versions: Dict[str, str],
+    ):
+        self.autodiscovery_manager = autodiscovery_manager
+        self.autodiscovered_cached_service = None
+        self.autodiscovered_extension_versions_hash = hash(
+            tuple(sorted(current_extension_versions.items()))
+        )
+
+        autodiscovery_query_interval = config.get_autodiscovery_querry_interval()
+        if autodiscovery_query_interval >= 60:
+            self.query_interval = timedelta(minutes=autodiscovery_query_interval)
+        else:
+            logging_context.log(
+                "Query interval for autodiscovery cannot be smaller than 60 minutes. Run with default 60 minutes interval"
+            )
+            self.query_interval = timedelta(minutes=60)
+
+        self.autodiscovery_task = asyncio.create_task(
+            self.autodiscovery_manager.get_autodiscovery_service(services)
+        )
+        self.time_since_last_autodiscovery = datetime.now()
+
+    async def get_task_result(self):
+        if self.autodiscovery_task is not None:
+            ad_service_result = await self.autodiscovery_task
+            if ad_service_result:
+                self.autodiscovered_cached_service = ad_service_result
+
+    async def _refresh_autodiscovery_task(
+        self, services, new_extension_versions_hash, force_refresh=False
+    ):
+        if not force_refresh and self.autodiscovery_task and not self.autodiscovery_task.done():
+            return
+
+        await self.get_task_result()
+
+        self.autodiscovery_task = asyncio.create_task(
+            self.autodiscovery_manager.get_autodiscovery_service(services)
+        )
+
+        self.autodiscovered_extension_versions_hash = new_extension_versions_hash
+        self.time_since_last_autodiscovery = datetime.now()
+
+    async def _handle_autodiscovery_task_result(self):
+        if self.autodiscovery_task:
+            ad_service_result = await self.autodiscovery_task
+            if ad_service_result:
+                self.autodiscovered_cached_service = ad_service_result
+            self.autodiscovery_task = None
+
+    async def get_cached_or_refreshed_metrics(
+        self, services: List[GCPService], new_extension_versions: Dict[str, str]
+    ):
+        new_extension_versions_hash = hash(tuple(sorted(new_extension_versions.items())))
+        delta = datetime.now() - self.time_since_last_autodiscovery
+
+        if self.autodiscovery_task is not None and self.autodiscovery_task.done():
+            await self._handle_autodiscovery_task_result()
+
+        if (
+            new_extension_versions_hash != self.autodiscovered_extension_versions_hash
+            or delta > self.query_interval
```

Comment:
Does it make sense to ask this if the hash is already equal? I mean, there's an OR, if delta is > than the interval but the hash didn't change, why to run the following task?